src/main/scala/cognite/spark/v1/SequenceRowsRelation.scala

```diff
@@ -0,0 +1,318 @@
+          .map {
+            case (_, rows) =>
+              rows.map { r =>
+                val values = expectedColumns
+                  .zip(r.values)
+                  .map {
+                    case (column, value) =>
+                      parseJsonValue(value, columnTypes(column))
+                        .getOrElse(throw new Exception(s"Unexpected value $value in column $column"))
+                  }
+                ProjectedSequenceRow(r.rowNumber, values)
+              }
+          }
+
+      fs2.Stream.eval(projectedRows).flatMap(r => r)
+    }
+
+  def fromRow(schema: StructType): (Array[String], Row => SequenceRow) = {
+    val rowNumberIndex = schema.fieldNames.indexOf("rowNumber")
+    if (rowNumberIndex < 0) {
+      throw new Exception("Can't upsert sequence rows, column `rowNumber` is missing.")
+    }
+
+    val columns = schema.fields.zipWithIndex.filter(_._1.name != "rowNumber").map {
+      case (field, index) =>
+        val seqColumn = sequenceInfo.columns
+          .find(_.externalId == field.name)
+          .getOrElse(throw new Exception(
+            s"Can't insert column `${field.name}` into sequence $sequenceId, the column does not exist in the sequence definition"))
+        def jsonFromDouble(num: Double): Json =
+          Json.fromDouble(num).getOrElse(throw new Exception(s"Numeric value $num"))
+        val tryGetValue: PartialFunction[Any, Json] = seqColumn.valueType match {
+          case "DOUBLE" => {
+            case null => Json.Null
+            case x: Double => jsonFromDouble(x)
+            case x: Int => jsonFromDouble(x.toDouble)
+            case x: Float => jsonFromDouble(x.toDouble)
+            case x: Long => jsonFromDouble(x.toDouble)
+            case x: java.math.BigDecimal => jsonFromDouble(x.doubleValue)
+            case x: java.math.BigInteger => jsonFromDouble(x.doubleValue)
+          }
+          case "LONG" => {
+            case null => Json.Null
+            case x: Int => Json.fromInt(x)
+            case x: Long => Json.fromLong(x)
+          }
+          case "STRING" => {
+            case null => Json.Null
+            case x: String => Json.fromString(x)
+          }
+        }
+        (index, seqColumn, tryGetValue)
+    }
+
+    def parseRow(row: Row): SequenceRow = {
+      val rowNumber = row.get(rowNumberIndex) match {
+        case x: Long => x
+        case x: Int => x: Long
+        case _ => throw SparkSchemaHelperRuntime.badRowError(row, "rowNumber", "Long", "")
+      }
+      val columnValues = columns.map {
+        case (index, seqColumn, tryGetValue) =>
+          tryGetValue.applyOrElse(
+            row.get(index),
+            (_: Any) =>
+              throw SparkSchemaHelperRuntime
+                .badRowError(row, seqColumn.externalId, seqColumn.valueType, "")
+          )
+      }
+      SequenceRow(rowNumber, columnValues)
+    }
+
+    (columns.map(_._2.externalId), parseRow)
+  }
+
+  def delete(rows: Seq[Row]): IO[Unit] = ???
```

Comment:
Woops? :) Can we support deletes in an easy way? Probably not yet (would be similar to data points deletes, I suppose), in which case it's fine to throw an exception about it not being supported.

Reply:
Sequence rows are deleted by the `rowNumber`, so I have just added that.